[SYCL][ESIMD] Add test for esimd fill constructor

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add test for simd fill constructor

The test was splitted into two files:
- core, that iterating only C++ with base data types
- fp_extra, that itarating with sycl::half data type

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+
+namespace esimd = sycl::ext::intel::experimental::esimd;
+namespace esimd_tests = esimd_test::api::functional::ctors;
+namespace esimd_functional = esimd_test::api::functional;
```
1) What these alias are for? Why don't you use them in `using` statements in cpp files then?
2) Is there a possibility of name clash with `sycl::ext::intel::experimental::esimd`?
3) Doesn't `esimd_tests` looks like `esimd_test` with the entirely different meaning behind?
4) Why all declarations below are in global namespace? For example, `::initializer` struct.

vasilytric replied:
I don't want to use `using namespace ...` statements in headers

yuriykoch replied:
We might move the `esimd` and `esimd_functional` aliases to the `functional/common.hpp`. It looks quite general. What do you think?

vasilytric replied:
This comment will be applied, when one of new tests will be merged. I am going to use this aliases in many tests

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(DataT init_value, DataT step, DataT *out) {
+    esimd::simd<DataT, NumElems> simd_by_init =
+        esimd::simd<DataT, NumElems>(init_value, step);
```
```suggestion
    const auto simd_by_init =
        esimd::simd<DataT, NumElems>(init_value, step);
```
Minor, we might just avoid duplicating the same statement, as the type is strictly defined by the initializer itself

vasilytric replied:
This comments was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R31-R32)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+
+  template <typename DataT, int NumElems>
+  static void call_simd_ctor(DataT init_value, DataT step, DataT *out) {
+    esimd::simd<DataT, NumElems> simd_by_var_decl(init_value, step);
```
```suggestion
    const esimd::simd<DataT, NumElems> simd_by_var_decl(init_value, step);
```
Here and elsewhere: we might enforce const-correctness to state explicitly that this object is not going to change after the constructor was called. It may not defend from the `mutable` attributes in simd vector implementation (if any), but it definitely will trigger a failure in case of modyfying `copy_to()` call.
Minor for this specific case, still it may be a good practice to enforce const-correctness over the code in general.

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R43-R44)

yuriykoch replied:
Would you mind to apply both the `const-correctness` and the `auto` usage comments in every context descriptor used?

vasilytric replied:
Could you specify where `const-correctness` not applied?

yuriykoch replied:
My bad, misread the code.

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+
+// Enumeration of possible values for base value and step that will be provided
+// into simd constructor.
+enum class init_val {
+  min,
```
```suggestion
enum class init_val {
  none = 0,
  min,
```
Minor, just in general.
By using `= 0` we are stating the values of enum explicitly.
By providing some `none` value we have some `default` value to use for variable initialization and further failure trigger in case of some error in algorithm.

vasilytric replied:
I going to use this enum only in comparative statements, so we don't care about values of the fields

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+
+// Constructing a value for step and base values that depends on input
+// parameters.
+template <typename DataT, init_val BaseVal> DataT get_base_value() {
```
Looks familiar a bit.. Why not just use some vector of reference data and iterate over it in runtime? Is it possible, or the code will bloat?

```suggestion
template <typename DataT, init_val BaseVal> DataT get_value() {
```
Nevertheless, looks like the `get_base_value` is used for both the `base` and `step` here)

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R104)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+    esimd_tests::shared_vector<DataT> result{NumElems,
+                                             shared_allocator<DataT>{queue}};
+    esimd_tests::shared_vector<DataT> shared_ref_data{
+        ref_data.begin(), ref_data.end(), shared_allocator<DataT>{queue}};
```
```suggestion
  auto
  call_simd(sycl::queue &queue, const std::vector<DataT> &ref_data) {
    using vector_t = esimd_tests::shared_vector<DataT>;
    vector_t result{NumElems, shared_allocator<DataT>{queue}};
    vector_t shared_ref_data{
        ref_data.begin(), ref_data.end(), shared_allocator<DataT>{queue}};
```
Minor - you might avoid repeating the same type

vasilytric replied:
not actual commit, now only one shared vector using

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+
+// Constructing verbose failure message.
+template <int NumElems>
+void on_failure(const std::string &step_value, const std::string &base_value,
```
1 - Aren't you enforced to have `std::to_string()` repeated later in the code by using std::string here? Looks like you might to pass the values actually, and then just provide the generic way to "stringify" them
2 - Is the floating point string representation good here? Is it possible to provide some generic way to print fp-values with hex values printed (nan opcode for example)?

vasilytric replied:
I've applied this commit [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R242-R244)

yuriykoch replied:
That's definitely better, still we need to look into the test sources to have a hint about actual values used. Are we going to add hex representation later?

vasilytric replied:
> still we need to look into the test sources to have a hint about actual values used

Do you mean comment about what happens in the test? Yes, I am going to add it later

> Are we going to add hex representation later?

Yes, we have a plans for change representation of floating point types for printing it values to the console

yuriykoch replied:
> Do you mean comment about what happens in the test?

Nope, I'm talking about console output on test failure. Still looks like you have a plenty of information printed already.

vasilytric replied:
console output will be looks like:
```
Failed for simd<unsigned short, 32>, with context: variable declaration, retrieved: 32767, expected: 65504, at index: 31, with base value: max_half, with step value: negative
```

@yuriykoch  Which information should be added?

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_fp_extra.cpp:
```diff
@@ -0,0 +1,46 @@
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu, level_zero
+// XREQUIRES: gpu
+// TODO gpu and level_zero in REQUIRES due to only this platforms supported yet.
+// The current "REQUIRES" should be replaced with "gpu" only as mentioned in
+// "XREQUIRES".
```
```suggestion
// TODO Remove the level_zero restriction once the test is supported on other platforms
```

vasilytric replied:
Now I have approved comment and this greater than suggested comment, but current comment provides more details about necessary actions

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_fp_extra.cpp:
```diff
@@ -0,0 +1,46 @@
+// XRUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: false
+// XFAIL: *
+// TODO The simd can't be constructed with sycl::half data type. The issue was
+// created (https://github.com/intel/llvm/issues/5077) and the this test must be
+// enabled when it is resolved.
```
```suggestion
// TODO The simd can't be constructed with the sycl::half data type.
// Enable the test after https://github.com/intel/llvm/issues/5077 resolved.
```
What do you think?

vasilytric replied:
Now I have approved comment and this greater than suggested comment, but current comment provides more details about necessary actions

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+      call_simd_struct<BaseVal, Step>{});
+
+  return pass;
+}
```
```suggestion
}

```

vasilytric replied:
done [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R274)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+// Iterating over provided types and dimensions, running test for each of them.
+template <typename TestT, init_val BaseVal, init_val Step, typename... Types,
+          int... Ints>
+bool run_verifying(sycl::queue &queue,
```
```suggestion
bool run_verification(sycl::queue &queue,
```
Aren't you using the `run_verification_with_chosen_test_type` name for the similar purpose in `ctor_default.cpp`?

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R203)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+
+// Iterating over provided types and dimensions, running test for each of them.
+template <typename TestT, init_val BaseVal, init_val Step, typename... Types,
+          int... Ints>
```
```suggestion
          int... Dims>
```

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R257)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+// equal to an expected value.
+template <init_val BaseVal, init_val Step> struct verify_obtained_result {
+  template <typename DataT, int NumElems>
+  bool verify_test_result(const std::vector<DataT> &ref_data,
```
Isn't `verify_obtained_result.verify_test_result()` a bit of duplication? May be something like `validator.verify()` will fit?
Generally, it's better to use nouns to name classes/instances and to use verbs to name methods/functions.

vasilytric replied:
the test structure was changed

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+
+  pass &= esimd_functional::for_all_types_and_dims<esimd_tests::test, TestT>(
+      types, dimensions, queue, verify_obtained_result<BaseVal, Step>{},
+      call_simd_struct<BaseVal, Step>{});
```
Wait a minute. It's quite hard to follow the call flow.
So we have:

- `verify_obtained_result` struct - to validate results
- `call_simd_struct` - to retrieve such results
- `TestT` which is basically the case of context usage
- some `for_all_types_and_dims` which seems to iterate over types and dimensions and call... some `esimd_tests::test`?

So.. Where can I find the root logic for the test, and do we really need to split it over multiple functors? What's the advantage to have a separate `verify_obtained_result` for example? Both the `call_simd_struct` and `verify_obtained_result` looks like very much fill-ctor-specific..

vasilytric replied:
the test structure was changed

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+                           std::to_string(result_data[zero_elem]), data_type,
+                           test_case_description);
+    }
+    for (size_t i = 1; i < ref_data.size(); ++i) {
```
```suggestion
    for (size_t i = 1; i < result_data.size(); ++i) {
```
or:

- move out the size() call according to the llvm code style
- use NumElems if appropriate, with preliminary assertions on vector sizes

vasilytric replied:
comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R228)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,226 @@
+    DataT base_value{get_base_value<DataT, BaseVal>()};
+    size_t zero_elem{0};
+
+    if (!are_bitwise_equal(ref_data[zero_elem], base_value)) {
```
Is `ref_data` a reference data? Why do we need to check the first item from the `ref_data`, the rest of items from the `result_data`?

vasilytric replied:
now we compare value from `result` vector with expected value, I've done it [here](https://github.com/intel/llvm-test-suite/pull/598/commits/652dcd3b7c72dceecf81bb7bd486398d8664c24d#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R232)

vasilytric committed:
Merge branch 'intel' into ctor_fill_b

vasilytric committed:
[SYCL][ESIMD] Structure changes in tests on simd fill constructor

- run_test and verification function defined in ctor_fill.hpp
- add class for structurize log error message
- rename some function and variables
- remove initialization

vasilytric committed:
[SYCL][ESIMD] Add test on simd vector constructor accuracy

- update ctor_fill.hpp
- rename some variables
- add test for vector ctor accuracy

vasilytric committed:
[SYCL][ESIMD] Remove interacting with reference data

Because it didn't used in this test

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,281 @@
+    return value<DataT>::denorm_min();
+  } else if constexpr (Value == init_val::inexact) {
+    return 0.1;
+  } else if constexpr (Value == init_val::ulp || Value == init_val::ulp_half) {
```
Would you mind to move the ulp value generation to a separate function?
For example, something like:
```
value<T>::pos_ulp(T base) {
  ulp(base, value<T>::inf());
}
value<T>::neg_ulp(T base) {
  ulp(base, -value<T>::inf());
}
```

yuriykoch replied:
Sorry, still applicable for c2320ccbc7baf5f52441fb4a3b711f4e0cc97ae0 - it's not great to have nextafter(base, 1.0) in general, and ulp might differ depending on direction.

vasilytric replied:
This comment was applied in this commit: ce131e3d6d50d19297f1bebe5b45adbd86e58347

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,281 @@
+
+    bool passed = true;
+
+    DataT expected_value = base_value;
+    for (size_t i = 0; i < result.size(); ++i) {
+      // std::isnan() couldn't be called for integral types because it call is
+      // ambiguous GitHub issue for that case:
+      // https://github.com/microsoft/STL/issues/519
+      if (!are_bitwise_equal(result[i], expected_value)) {
+        if constexpr (type_traits::is_sycl_floating_point_v<DataT>) {
+          if (std::isnan(result[i]) &&
+              (std::isnan(step_value) || std::isnan(base_value))) {
+            continue;
+          }
+        }
+        passed = false;
```
```suggestion
    if constexpr ((Base == init_val::nan) || (Step == init_val::nan)) {
      if (!are_bitwise_equal(result[0], base_value))) {
        ...
      }
      for (size_t i = 1; i < result.size(); ++i) {
        if (!isnan(result[i])) {
          ...
        }
      }
    } else {
      DataT expected_value = base_value;
      for (size_t i = 0; i < result.size(); ++i) {
        if (!are_bitwise_equal(result[i], expected_value)) {
          passed = false;
```
with
```
  constexpr Base = BaseValT::value;
  constexpr Step = StepT::value;
```

Otherwise we would miss the difference in opcode for `result[0]`

vasilytric replied:
This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/209a007ff239c59795d58172a05bf77ba4fd5b80#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4L224-L244)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,281 @@
+        passed = false;
+
+        const auto description =
+            FillCtorTestDescription<DataT, NumElems, TestCaseT, BaseVal::value,
+                                    Step::value>(i, result[i - 1] + step_value,
+                                                 expected_value, data_type);
```
```suggestion
            FillCtorTestDescription<DataT, NumElems, TestCaseT, BaseVal::value,
                                    Step::value>(i, retrieved_value,
                                                 expected_value, data_type);
```
with
```
const auto& retrieved = result[i];
```

vasilytric replied:
Now `result[i]` provided into `FillCtorTestDescription` I've done it in this commit: 209a007ff239c59795d58172a05bf77ba4fd5b80

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_accuracy.cpp:
```diff
@@ -0,0 +1,56 @@
+
+  std::cout << (passed ? "=== Test passed\n" : "=== Test FAILED\n");
+  return passed ? 0 : 1;
+}
```
```suggestion
}

```

vasilytric replied:
I've applied this comment in this commit: f54e62c35c53b1c5ef38651865415b7023d08bb6

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_accuracy.cpp:
```diff
@@ -0,0 +1,56 @@
+  const auto fp_types = get_tested_types<tested_types::fp>();
+  const auto single_dim = values_pack<8>();
+
+  passed &= run_verification<var_dec, init_val::denorm, init_val::ulp>(
```
It's unclear what is base and what is step. Would you mind to add a comment above in the both cpp files?

vasilytric replied:
I've added comment that described where defined step and where defined base [here](https://github.com/intel/llvm-test-suite/pull/598/commits/1c807fc362443314f1e18c19bc469cffaef58e2b#diff-363883cae3ba91f2ba9d07628c42317d4da5a038cb6bef71d8f3023281b6cc79R41-R42)

vasilytric committed:
[SYCL][ESIMD] Move function for receiving ulp value into value.hpp

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_core.cpp:
```diff
@@ -0,0 +1,116 @@
+using namespace esimd_test::api::functional::ctors;
+using namespace esimd_test::api::functional;
+
+int main(int argc, char **argv) {
```
```suggestion
int main(int, char **) {
```

vasilytric replied:
I've removed it [here](https://github.com/intel/llvm-test-suite/pull/598/commits/1c807fc362443314f1e18c19bc469cffaef58e2b#diff-363883cae3ba91f2ba9d07628c42317d4da5a038cb6bef71d8f3023281b6cc79L30)

vasilytric committed:
[SYCL][ESIMD] Change "min" to "lowest" in init_val_to_string function

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_core.cpp:
```diff
@@ -0,0 +1,116 @@
+
+  bool passed = true;
+
+  const auto one_and_eighth_dims = values_pack<1, 8>();
```
Looks like that if we change to the `value_pack<1, 16>` for some reason it would result in huge renaming and reformatting below. May be something less specific? May be something like:
```
const auto all_dims = values_pack<1, 8, 16, 32>();
const auto specific_dims = values_pack<1, 8>();
```
or
```
const auto all_dims = values_pack<1, 8, 16, 32>();
const auto two_dims = values_pack<1, 8>();
const auto single_dim = values_pack<8>();
```
depending on the logic behind such selection. What would you say? Same for the `char_int_types` maybe.

vasilytric replied:
I've applied it [here](https://github.com/intel/llvm-test-suite/pull/598/commits/1c807fc362443314f1e18c19bc469cffaef58e2b#diff-6df0727accdf4d3a754136946c897c835113f6e13a4439c34444a0afed9eb945L36-L37)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_fp_extra.cpp:
```diff
@@ -0,0 +1,50 @@
+// RUN: false
+// XFAIL: *
+// TODO The simd can't be constructed with sycl::half data type. The issue was
+// created (https://github.com/intel/llvm/issues/5077) and the this test must be
```
```suggestion
// created (https://github.com/intel/llvm/issues/5077) and this test must be
```

vasilytric replied:
may be this statement will be better `and the test must be`?
I've use it [here](https://github.com/intel/llvm-test-suite/pull/598/commits/1c807fc362443314f1e18c19bc469cffaef58e2b#diff-363883cae3ba91f2ba9d07628c42317d4da5a038cb6bef71d8f3023281b6cc79L19-L20)

yuriykoch replied:
Good for me either way, thanks)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_accuracy.cpp:
```diff
@@ -0,0 +1,56 @@
+
+  bool passed = true;
+
+  const auto fp_types = get_tested_types<tested_types::fp>();
```
Should there be any `core` or `fp_extra` suffix for this file?

vasilytric replied:
I think that is not necessary because this file will be splitted into `core` and `fp_extra` files. I've done it in this commit: f54e62c35c53b1c5ef38651865415b7023d08bb6

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,281 @@
+    } else if constexpr (Val == init_val::ulp_half) {
+      return "ulp_half";
+    } else {
+      static_assert(Val != Val, "Unexpected base value value");
```
Could it possibly trigger for step, not base? Or am I missing something?

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/598/commits/209a007ff239c59795d58172a05bf77ba4fd5b80#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4L190-L191)

vasilytric committed:
[SYCL][ESIMD] Updates in ctor_fill.hpp

- Change logic forverification if base or step value is nan and all
  other cases
- Update static_assert messages

vasilytric committed:
[SYCL][ESIMD] Add pos_ulp and neg_ulp into values

Use value::pos_ulp() in ctor fill test instead of using value::ulp()

vasilytric committed:
[SYCL][ESIMD] Small updates in ctor fill tests

- Rename values_pack with two dims
- Update comments
- Remove unused args

vasilytric committed:
[SYCL][ESIMD] Split ctor_fill_accuracy test into core and fp extra

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -219,32 +219,45 @@ class run_test {
+private:
+  bool m_passed = true;
+
+  void fail_test(size_t step, DataT retrieved, DataT expected,
```
```suggestion
  void fail_test(size_t index, DataT retrieved, DataT expected,
```
The `base` and `step` has specific meaning for fill constructor.

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/04d6bab9c3fb63aeef04dbeea45f802d05edce82#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4L258-L259)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -219,32 +219,45 @@ class run_test {
+  }
+
+private:
+  bool m_passed = true;
```
De-facto it's still the result of `operator()`, so you need to initialize it every time you enter the `operator()`, otherwise you might end up with wrong result after 2 sequential calls to `operator()` for the same instance.

Also what would you say about something like `passed = fail_test(...);`?

vasilytric replied:
this comment was applied [here]()

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -219,32 +219,45 @@ class run_test {
     queue.wait_and_throw();
 
-    bool passed = true;
+    if (!are_bitwise_equal(result[0], base_value)) {
```
```suggestion
    // Verify the base value was passed as-is
    if (!are_bitwise_equal(result[0], base_value)) {
```

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/04d6bab9c3fb63aeef04dbeea45f802d05edce82#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R222)

yuriykoch replied:
Minor, still looks a bit confusing to me:
```
if (!are_bitwise_equal(...)) {
  //  Verify ...
  passed = fail_test(...);
}

init_for_verification;
// Verify ...
for (..) {
  check_something;
}
```
Doesn't such comment placement mean the actual verification is made by `fail_test()`, not by `are_bitwise_equal()`? What would you say about using something like this instead?

```
//  Verify ...
if (!are_bitwise_equal(...)) {
  passed = fail_test(...);
}

// Verify ...
init_for_verification;
for (..) {
  check_something;
}
```
By doing so you might provide a comment to the block of code, visually and semantically separating them

yuriykoch replied:
Minor, still looks a bit confusing to me:
```
if (!are_bitwise_equal(...)) {
  //  Verify ...
  passed = fail_test(...);
}

init_for_verification;
// Verify ...
for (..) {
  check_something;
}
```
Doesn't such comment placement mean the actual verification is made by `fail_test()`, not by `are_bitwise_equal()`? What would you say about using something like this instead?

```
//  Verify ...
if (!are_bitwise_equal(...)) {
  passed = fail_test(...);
}

// Verify ...
init_for_verification;
for (..) {
  check_something;
}
```
By doing so you might provide a comment to the block of code, visually and semantically separating them

vasilytric replied:
I've applied this comment in this commit: f4636f5c6bbda87420b54c9ceb07763559d30865

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -219,32 +219,45 @@ class run_test {
+        expected_value += step_value;
+      }
+    } else {
+      for (size_t i = 1; i < result.size(); ++i) {
```
You've changed the scope of iteration. Please, change the logic of `expected` accordingly.

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/04d6bab9c3fb63aeef04dbeea45f802d05edce82#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R249-R252)
now `expected_value` will be increased before verification

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -219,32 +219,45 @@ class run_test {
-          if (std::isnan(result[i]) &&
-              (std::isnan(step_value) || std::isnan(base_value))) {
-            continue;
+    if constexpr (BaseVal::value == init_val::nan ||
```
```suggestion
    // Verify the step value works as expected being passed to the fill constructor
    if constexpr (BaseVal::value == init_val::nan ||
```

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/04d6bab9c3fb63aeef04dbeea45f802d05edce82#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4R227-R228)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_accuracy_fp_extra.cpp:
```diff
@@ -0,0 +1,61 @@
+#include "ctor_fill.hpp"
+
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional::ctors;
+using namespace esimd_test::api::functional;
```
```suggestion
using namespace esimd_test::api::functional;
```
What do you think?

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/04d6bab9c3fb63aeef04dbeea45f802d05edce82#diff-7bc755870d95f59f1370a04cf56a312662d85c6d91a6e6176d4537f89e4e304cL30) and in other files

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -219,32 +219,45 @@ class run_test {
+        if (!std::isnan(result[i])) {
+          if (!are_bitwise_equal(result[i], expected_value)) {
+            fail_test(i, result[i], expected_value, data_type);
           }
```
Please note that there is no need to check for bitwise equality. It's already known the values are different: expected NaN, retrieved non-NaN

vasilytric replied:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/598/commits/04d6bab9c3fb63aeef04dbeea45f802d05edce82#diff-7fc90b589428e6e89076f07f5ff280a217b7f5c8e836bfe6c742734825074ef4L231-L234)

vasilytric committed:
[SYCL][ESIMD] Update tests for fill ctor

- Remove bitwise comparison in case step or base value is nan
- Remove unnecessary using namespace ...
- Update comments

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_fill.hpp:
```diff
@@ -0,0 +1,292 @@
+        if (!std::isnan(result[i])) {
+          passed = false;
+
+          // TODO: Make ITestDescription architecture more flexible
+          std::string log_msg = "Failed for simd<";
```
```suggestion
          // TODO: Make ITestDescription architecture more flexible
          // We are assuming that the NaN opcode may differ
          std::string log_msg = "Failed for simd<";
```

vasilytric replied:
this comment was applied in this commit: 292118ac7aa44deec887e1fb9b977fa49e636820

vasilytric committed:
[SYCL][ESIMD] Change comments structure

Move comments before code that it describe

vasilytric committed:
[SYCL][ESIMD] Explicitly show which uses from ctors namespace

vasilytric committed:
[SYCL][ESIMD] Update comments to ctor fill tests

vasilytric committed:
[SYCL][ESIMD] Add the link to open source issue to test disable reason

vasilytric committed:
[SYCL][ESIMD] Update TODO comment

vasilytric committed:
[SYCL][ESIMD] Update comments

vasilytric commented:
@v-klochkov Vyacheslav, could you review this PR?

v-klochkov commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -112,6 +112,19 @@ template <typename DataT> struct value {
+  static DataT ulp(DataT base_val, DataT direction) {
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return static_cast<sycl::half>(
+          (sycl::nextafter(base_val, direction) - base_val) * 8192);
```
Would you please add a comment here. Why 8192 is used for sycl::half ?

vasilytric replied:
comment was added in this commit: a756ab394043ea46d847f0e2d30d7139c13eecb2

v-klochkov commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -112,6 +112,19 @@ template <typename DataT> struct value {
+
+  static DataT pos_ulp(DataT base_val) { return ulp(base_val, inf()); }
+
+  static DataT neg_ulp(DataT base_val) { return ulp(base_val, -inf()); }
```
This func is unused. Shouldn't it be either deleted or used?

vasilytric replied:
This function will be useful in the next tests

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_accuracy_fp_extra.cpp:
```diff
@@ -0,0 +1,65 @@
+// XRUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: false
+// XFAIL: *
+// TODO The simd can't be constructed with sycl::half data type. The issue was
```
Please try the newer compiler. sycl::half should be working fine with simd now.

vasilytric replied:
This discuss related to all tests, so suggest discuss it in the one place, in this your comment: [discussion in PR#687](https://github.com/intel/llvm-test-suite/pull/687#discussion_r778299304)

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_accuracy_fp_extra.cpp:
```diff
@@ -0,0 +1,65 @@
+
+  bool passed = true;
+
+  const auto fp_types = get_tested_types<tested_types::fp_extra>();
```
The types like sycl::half need runtime checks verifying that device supports them. E.g. sycl::aspect::fp16

vasilytric replied:
`fp_extra` files was temporary removed and the `sycl::half` data type was enable for all tests, I've done it this commit: 716223f66915597c8e6b289752ed0e25b3eb6372

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_core.cpp:
```diff
@@ -0,0 +1,154 @@
+// XRUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: false
+// XFAIL: *
+// TODO This test disabled due to simd<short, 32> vector filled with unexpected
```
Rather than disabling the whole test, it is better to temporarily disable the 32 element vectors in this test.
For example, create a macro: "#define SKIP_VECTOR_LEN_32" and add a comment with request to eliminate the macro when the problem/issue is fixed.
My other point is that the test is not even compiled now, and thus it is unclear if it is compilable, if it has syntax errors in it, etc.

vasilytric replied:
I've added a macros to disable some tests in this commit: 716223f66915597c8e6b289752ed0e25b3eb6372

vasilytric committed:
[SYCL][ESIMD] Enable tests for CI and temporary remove fp_extra tests

vasilytric committed:
[SYCL][ESIMD] Add comment to value::ulp() function

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_core.cpp:
```diff
@@ -90,7 +89,7 @@ int main(int, char **) {
                                     ctors::init_val::positive>(queue, two_dims,
                                                                char_int_types);
-
+#ifdef SIMD_RUN_TEST_WITH_VECTOR_LEN_32
   const auto all_dims = values_pack<1, 8, 16, 32>();
```
Shouldn't it be something like this?:
```suggestion
#ifdef SIMD_RUN_TEST_WITH_VECTOR_LEN_32
  const auto all_dims = values_pack<1, 8, 16, 32>();
#else
  const auto all_dims = values_pack<1, 8, 16>();
#endif
```

vasilytric replied:
I've applied this comment in this commit: a3b04bede68dfb78f3144647d94e7847fa385525

v-klochkov commented on SYCL/ESIMD/api/functional/ctors/ctor_fill_core.cpp:
```diff
@@ -0,0 +1,153 @@
+  passed &= ctors::run_verification<ctors::const_ref, ctors::init_val::min_half,
+                                    ctors::init_val::positive>(queue, two_dims,
+                                                               char_int_types);
+#ifdef SIMD_RUN_TEST_WITH_VECTOR_LEN_32
+  const auto all_dims = values_pack<1, 8, 16, 32>();
```
Sorry, if I duplicate this comment. The first one disappeared, I don't see it.
Wouldn't it be better way to say something like this:
```suggestion
#ifdef SIMD_RUN_TEST_WITH_VECTOR_LEN_32
  const auto all_dims = values_pack<1, 8, 16, 32>();
#else
  const auto all_dims = values_pack<1, 8, 16>();
#endif  
```

vasilytric replied:
I've applied this comment in this commit: a3b04bede68dfb78f3144647d94e7847fa385525

vasilytric committed:
[SYCL][ESIMD] Remove 32 vector lent in case macros is not defined

vasilytric commented:
@v-klochkov all CI verifications are passed. Can we merge this PR?
